spring-rabbit/src/main/java/org/springframework/amqp/rabbit/core/RabbitTemplate.java

```diff
@@ -1627,25 +1627,23 @@ private Message doSendAndReceiveWithDirect(String exchange, String routingKey, M
 		if (this.usePublisherConnection && connectionFactory.getPublisherConnectionFactory() != null) {
 			connectionFactory = connectionFactory.getPublisherConnectionFactory();
 		}
-		DirectReplyToMessageListenerContainer container = this.directReplyToContainers.get(connectionFactory);
-		if (container == null) {
-			synchronized (this.directReplyToContainers) {
-				container = this.directReplyToContainers.get(connectionFactory);
-				if (container == null) {
-					container = new DirectReplyToMessageListenerContainer(connectionFactory);
-					container.setMessageListener(this);
-					container.setBeanName(this.beanName + "#" + this.containerInstance.getAndIncrement());
-					if (this.taskExecutor != null) {
-						container.setTaskExecutor(this.taskExecutor);
-					}
-					if (this.afterReceivePostProcessors != null) {
-						container.setAfterReceivePostProcessors(this.afterReceivePostProcessors
-								.toArray(new MessagePostProcessor[this.afterReceivePostProcessors.size()]));
-					}
-					container.start();
-					this.directReplyToContainers.put(connectionFactory, container);
-					this.replyAddress = Address.AMQ_RABBITMQ_REPLY_TO;
+		DirectReplyToMessageListenerContainer container;
```

Comment:
Would you mind explaining why this change, please?

The idea is like try to get `container` from the map concurrently without any locks.
And only if it is null, go to the `synchronized (this.directReplyToContainers)` and try to do that again.

But now, with your, change we always block all other threads because the first one has decided to go to the `synchronized (this.directReplyToContainers)` block. The problem is going to be exponential for all the waiting threads and actually for nothing, if we have already populated `container` before.

We borrowed this style from the Singleton Pattern implementation: https://www.journaldev.com/1377/java-singleton-design-pattern-best-practices-examples#thread-safe-singleton. This is called "double checked locking principle is used".

What is idea do not do that?

Thanks

Reply:
I understand that this is an attempt to go with fastpath first, but the reasoning behind removing this is that underlying map implementation is not thread safe and so every access (read and write) should be done behind the lock, like in other places, otherwise we could see stale values. Basically it is the same as if we would synchronize setter but not a getter.

Reply:
Yeah... That's good point. 
But again: for consistency with the approach.

However in our case we never see stale value, because we never override it.
Only what we do is cleaning the map during `stop()`.

Does it make sense to keep "unsafe" `get()` ?

Thanks